### ISSUE TITLE
Observed in debug level message "Failed to load netty-tcnative; OpenSslEngine will be unavailable"

### ISSUE THREAD

**Issue by jiec-msft** [microsoft/ApplicationInsights-Java#2640]
With java agent version 3.3.1, will below errors (though at debug & warn level) brings any impacts?

```
2022-10-26 15:13:06.255Z DEBUG io.netty.handler.ssl.OpenSsl - Failed to load netty-tcnative; OpenSslEngine will be unavailable, unless the application has already loaded the symbols by some other means. See https://netty.io/wiki/forked-tomcat-native.html for more information.
java.lang.IllegalArgumentException: Failed to load any of the given libraries: [netty_tcnative_linux_x86_64, netty_tcnative_linux_x86_64_fedora, netty_tcnative_x86_64, netty_tcnative]
	at io.netty.util.internal.NativeLibraryLoader.loadFirstAvailable(NativeLibraryLoader.java:114)
	at io.netty.handler.ssl.OpenSsl.loadTcNative(OpenSsl.java:705)
	at io.netty.handler.ssl.OpenSsl.<clinit>(OpenSsl.java:146)
	at io.netty.handler.ssl.SslProvider.isAlpnSupported(SslProvider.java:53)
	at reactor.netty.http.Http2SslContextSpec.lambda$static$0(Http2SslContextSpec.java:170)
	at reactor.netty.tcp.AbstractProtocolSslContextSpec.configure(AbstractProtocolSslContextSpec.java:47)
	at reactor.netty.tcp.AbstractProtocolSslContextSpec.<init>(AbstractProtocolSslContextSpec.java:39)
	at reactor.netty.http.Http2SslContextSpec.<init>(Http2SslContextSpec.java:155)
	at reactor.netty.http.Http2SslContextSpec.forClient(Http2SslContextSpec.java:59)
	at reactor.netty.http.client.HttpClientSecure.<clinit>(HttpClientSecure.java:58)
	at reactor.netty.http.client.HttpClientConnect$MonoHttpConnect.lambda$subscribe$0(HttpClientConnect.java:220)
	at reactor.core.publisher.MonoCreate.subscribe(MonoCreate.java:58)
	at reactor.core.publisher.FluxRetryWhen.subscribe(FluxRetryWhen.java:77)
	at reactor.core.publisher.MonoRetryWhen.subscribeOrReturn(MonoRetryWhen.java:46)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:57)
	at reactor.netty.http.client.HttpClientConnect$MonoHttpConnect.subscribe(HttpClientConnect.java:272)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.FluxRetryWhen.subscribe(FluxRetryWhen.java:77)
	at reactor.core.publisher.MonoRetryWhen.subscribeOrReturn(MonoRetryWhen.java:46)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4385)
	at reactor.core.publisher.Mono.block(Mono.java:1706)
	at com.microsoft.applicationinsights.agent.internal.init.AppIdSupplier$GetAppIdTask.run(AppIdSupplier.java:139)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
	Suppressed: java.lang.UnsatisfiedLinkError: could not load a native library: netty_tcnative_linux_x86_64
		at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:224)
		at io.netty.util.internal.NativeLibraryLoader.loadFirstAvailable(NativeLibraryLoader.java:105)
		... 28 common frames omitted
	Caused by: java.io.FileNotFoundException: META-INF/native/libnetty_tcnative_linux_x86_64.so
		at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:171)
		... 29 common frames omitted
		Suppressed: java.lang.UnsatisfiedLinkError: no netty_tcnative_linux_x86_64 in java.library.path
			at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1860)
			at java.lang.Runtime.loadLibrary0(Runtime.java:871)
			at java.lang.System.loadLibrary(System.java:1124)
			at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:38)
			at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:376)
			at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:146)
			... 29 common frames omitted
			Suppressed: java.lang.UnsatisfiedLinkError: no netty_tcnative_linux_x86_64 in java.library.path
				at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1860)
				at java.lang.Runtime.loadLibrary0(Runtime.java:871)
				at java.lang.System.loadLibrary(System.java:1124)
				at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:38)
				at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
				at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
				at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
				at java.lang.reflect.Method.invoke(Method.java:498)
				at io.netty.util.internal.NativeLibraryLoader$1.run(NativeLibraryLoader.java:410)
				at java.security.AccessController.doPrivileged(Native Method)
				at io.netty.util.internal.NativeLibraryLoader.loadLibraryByHelper(NativeLibraryLoader.java:402)
				at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:368)
				... 30 common frames omitted
	Suppressed: java.lang.UnsatisfiedLinkError: could not load a native library: netty_tcnative_linux_x86_64_fedora
		at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:224)
		at io.netty.util.internal.NativeLibraryLoader.loadFirstAvailable(NativeLibraryLoader.java:105)
		... 28 common frames omitted
	Caused by: java.io.FileNotFoundException: META-INF/native/libnetty_tcnative_linux_x86_64_fedora.so
		at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:171)
		... 29 common frames omitted
		Suppressed: java.lang.UnsatisfiedLinkError: no netty_tcnative_linux_x86_64_fedora in java.library.path
			at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1860)
			at java.lang.Runtime.loadLibrary0(Runtime.java:871)
			at java.lang.System.loadLibrary(System.java:1124)
			at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:38)
			at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:376)
			at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:146)
			... 29 common frames omitted
			Suppressed: java.lang.UnsatisfiedLinkError: no netty_tcnative_linux_x86_64_fedora in java.library.path
				at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1860)
				at java.lang.Runtime.loadLibrary0(Runtime.java:871)
				at java.lang.System.loadLibrary(System.java:1124)
				at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:38)
				at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
				at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
				at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
				at java.lang.reflect.Method.invoke(Method.java:498)
				at io.netty.util.internal.NativeLibraryLoader$1.run(NativeLibraryLoader.java:410)
				at java.security.AccessController.doPrivileged(Native Method)
				at io.netty.util.internal.NativeLibraryLoader.loadLibraryByHelper(NativeLibraryLoader.java:402)
				at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:368)
				... 30 common frames omitted
	Suppressed: java.lang.UnsatisfiedLinkError: could not load a native library: netty_tcnative_x86_64
		at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:224)
		at io.netty.util.internal.NativeLibraryLoader.loadFirstAvailable(NativeLibraryLoader.java:105)
		... 28 common frames omitted
	Caused by: java.io.FileNotFoundException: META-INF/native/libnetty_tcnative_x86_64.so
		at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:171)
		... 29 common frames omitted
		Suppressed: java.lang.UnsatisfiedLinkError: no netty_tcnative_x86_64 in java.library.path
			at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1860)
			at java.lang.Runtime.loadLibrary0(Runtime.java:871)
			at java.lang.System.loadLibrary(System.java:1124)
			at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:38)
			at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:376)
			at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:146)
			... 29 common frames omitted
			Suppressed: java.lang.UnsatisfiedLinkError: no netty_tcnative_x86_64 in java.library.path
				at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1860)
				at java.lang.Runtime.loadLibrary0(Runtime.java:871)
				at java.lang.System.loadLibrary(System.java:1124)
				at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:38)
				at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
				at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
				at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
				at java.lang.reflect.Method.invoke(Method.java:498)
				at io.netty.util.internal.NativeLibraryLoader$1.run(NativeLibraryLoader.java:410)
				at java.security.AccessController.doPrivileged(Native Method)
				at io.netty.util.internal.NativeLibraryLoader.loadLibraryByHelper(NativeLibraryLoader.java:402)
				at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:368)
				... 30 common frames omitted
	Suppressed: java.lang.UnsatisfiedLinkError: could not load a native library: netty_tcnative
		at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:224)
		at io.netty.util.internal.NativeLibraryLoader.loadFirstAvailable(NativeLibraryLoader.java:105)
		... 28 common frames omitted
	Caused by: java.io.FileNotFoundException: META-INF/native/libnetty_tcnative.so
		at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:171)
		... 29 common frames omitted
		Suppressed: java.lang.UnsatisfiedLinkError: no netty_tcnative in java.library.path
			at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1860)
			at java.lang.Runtime.loadLibrary0(Runtime.java:871)
			at java.lang.System.loadLibrary(System.java:1124)
			at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:38)
			at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:376)
			at io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:146)
			... 29 common frames omitted
			Suppressed: java.lang.UnsatisfiedLinkError: no netty_tcnative in java.library.path
				at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1860)
				at java.lang.Runtime.loadLibrary0(Runtime.java:871)
				at java.lang.System.loadLibrary(System.java:1124)
				at io.netty.util.internal.NativeLibraryUtil.loadLibrary(NativeLibraryUtil.java:38)
				at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
				at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
				at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
				at java.lang.reflect.Method.invoke(Method.java:498)
				at io.netty.util.internal.NativeLibraryLoader$1.run(NativeLibraryLoader.java:410)
				at java.security.AccessController.doPrivileged(Native Method)
				at io.netty.util.internal.NativeLibraryLoader.loadLibraryByHelper(NativeLibraryLoader.java:402)
				at io.netty.util.internal.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:368)
				... 30 common frames omitted
2022-10-26 15:13:06.256Z DEBUG io.netty.handler.ssl.OpenSsl - Initialize netty-tcnative using engine: 'default'
2022-10-26 15:13:06.259Z DEBUG io.netty.handler.ssl.OpenSsl - Failed to initialize netty-tcnative; OpenSslEngine will be unavailable. See https://netty.io/wiki/forked-tomcat-native.html for more information.
java.lang.UnsatisfiedLinkError: io.netty.internal.tcnative.Library.aprMajorVersion()I
	at io.netty.internal.tcnative.Library.aprMajorVersion(Native Method)
	at io.netty.internal.tcnative.Library.initialize(Library.java:191)
	at io.netty.handler.ssl.OpenSsl.initializeTcNative(OpenSsl.java:710)
	at io.netty.handler.ssl.OpenSsl.<clinit>(OpenSsl.java:163)
	at io.netty.handler.ssl.SslProvider.isAlpnSupported(SslProvider.java:53)
	at reactor.netty.http.Http2SslContextSpec.lambda$static$0(Http2SslContextSpec.java:170)
	at reactor.netty.tcp.AbstractProtocolSslContextSpec.configure(AbstractProtocolSslContextSpec.java:47)
	at reactor.netty.tcp.AbstractProtocolSslContextSpec.<init>(AbstractProtocolSslContextSpec.java:39)
	at reactor.netty.http.Http2SslContextSpec.<init>(Http2SslContextSpec.java:155)
	at reactor.netty.http.Http2SslContextSpec.forClient(Http2SslContextSpec.java:59)
	at reactor.netty.http.client.HttpClientSecure.<clinit>(HttpClientSecure.java:58)
	at reactor.netty.http.client.HttpClientConnect$MonoHttpConnect.lambda$subscribe$0(HttpClientConnect.java:220)
	at reactor.core.publisher.MonoCreate.subscribe(MonoCreate.java:58)
	at reactor.core.publisher.FluxRetryWhen.subscribe(FluxRetryWhen.java:77)
	at reactor.core.publisher.MonoRetryWhen.subscribeOrReturn(MonoRetryWhen.java:46)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:57)
	at reactor.netty.http.client.HttpClientConnect$MonoHttpConnect.subscribe(HttpClientConnect.java:272)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.FluxRetryWhen.subscribe(FluxRetryWhen.java:77)
	at reactor.core.publisher.MonoRetryWhen.subscribeOrReturn(MonoRetryWhen.java:46)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4385)
	at reactor.core.publisher.Mono.block(Mono.java:1706)
	at com.microsoft.applicationinsights.agent.internal.init.AppIdSupplier$GetAppIdTask.run(AppIdSupplier.java:139)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
2022-10-26 15:13:07.338Z DEBUG io.netty.handler.ssl.JdkSslContext - Default protocols (JDK): [TLSv1.2] 
2022-10-26 15:13:07.338Z DEBUG io.netty.handler.ssl.JdkSslContext - Default cipher suites (JDK): [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_256_CBC_SHA, TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384]
```

```
2022-10-26 15:13:34.162Z WARN  c.a.m.o.e.i.p.TelemetryPipeline - Sending telemetry to the ingestion service: handshake timed out after 10000ms (https://eastus-3.in.applicationinsights.azure.com/v2.1/track) (future warnings will be aggregated and logged once every 5 minutes)
io.netty.handler.ssl.SslHandshakeTimeoutException: handshake timed out after 10000ms
	at io.netty.handler.ssl.SslHandler$7.run(SslHandler.java:2113)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:391)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:750)
2022-10-26 15:13:34.233Z WARN  r.n.http.client.HttpClientConnect - [ab63fd69, L:/10.244.4.12:44224 - R:eastus-3.in.applicationinsights.azure.com/40.71.13.169:443] The connection observed an error
io.netty.handler.ssl.SslHandshakeTimeoutException: handshake timed out after 10000ms
	at io.netty.handler.ssl.SslHandler$7.run(SslHandler.java:2113)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:391)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:750)
```

## Full log file

The logs start with "Start to read output" or "Start to read errors" is generated by my java app executing bash command "curl -iv xxxx".

[logs.app-with-tls-on-default-15-699b85995f-cjqj8-20221026.log](https://github.com/microsoft/ApplicationInsights-Java/files/9871153/logs.app-with-tls-on-default-15-699b85995f-cjqj8-20221026.log)

**Comment by trask** [comment]
hi @jiec-msft, this should be resolved in 3.4.0 and later (underlying issue was https://github.com/netty/netty/issues/12343)

**Comment by jiec-msft** [comment]
Thank you @Trask ***@***.***>. We use this agent for various customers. Could you help to describe more detailed impacts? Based on impacts, we’ll then adjust priority of upgrade  for customers.

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Trask Stalnaker ***@***.***>
发送时间: Thursday, October 27, 2022 12:20:04 AM
收件人: microsoft/ApplicationInsights-Java ***@***.***>
抄送: Mason Chen ***@***.***>; Mention ***@***.***>
主题: Re: [microsoft/ApplicationInsights-Java] Observed in debug level message "Failed to load netty-tcnative; OpenSslEngine will be unavailable" (Issue #2640)


hi @jiec-msft<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjiec-msft&data=05%7C01%7Cjiec%40microsoft.com%7Cd6bf0a59b3cd416f513708dab76df114%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638023980096292919%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=b5%2F176LLZbwx0ucn89GD0IeV9zx3SU6puvD8sghf13w%3D&reserved=0>, this should be resolved in 3.4.0 and later (underlying issue was netty/netty#12343<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnetty%2Fnetty%2Fissues%2F12343&data=05%7C01%7Cjiec%40microsoft.com%7Cd6bf0a59b3cd416f513708dab76df114%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638023980096449573%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=XpcfMAc3nm7FNwyubw4XC9poQiy8x4mhfZlxTIUwZU4%3D&reserved=0>)

―

**Comment by trask** [comment]
hi @jiec-msft, please see the netty issue linked above for details.

**Comment by jiec-msft** [comment]
Hi @trask , looking at the netty issue, my understand is it will always fail. But from application log, it's transmitted error (like below error message). So I'm wondering if netty issue is the root cause? If that's the cause, why it's causing transmitted issue in application insights rather than always failure? And will there be data log due to transmitted issue?

```
c.a.m.o.e.i.q.QuickPulsePingSender - In the last 5 minutes, the following operation has failed 3 times (out of 10): Pinging live metrics endpoint:
* io.netty.handler.ssl.SslHandshakeTimeoutException: handshake timed out after 10000ms
* 
```

**Comment by trask** [comment]
> Hi @trask , looking at the netty issue, my understand is it will always fail. But from application log, it's transmitted error (like below error message).

I'm sorry, I'm not following why these two things are not compatible, can you expand more to help me understand your thinking?

> If that's the cause, why it's causing transmitted issue in application insights rather than always failure?

Same here, I'm not following. Application Insights uses netty for transmitting data to the ingestion service, so it's expected that netty bug would cause transmission issues.

> And will there be data log due to transmitted issue?

I'm not sure what you mean by "data log", can you expand a bit?

**Comment by jiec-msft** [comment]
> Hi @trask , looking at the netty issue, my understand is it will always fail. But from application log, it's intermittent error (according to below error message). So I'm wondering if netty issue is the root cause? If that's the cause, why it's causing intermittent issue in application insights rather than always failure? ~~And will there be data log due to transmitted issue?~~
> 
> ```
> c.a.m.o.e.i.q.QuickPulsePingSender - In the last 5 minutes, the following operation has failed 3 times (out of 10): Pinging live metrics endpoint:
> * io.netty.handler.ssl.SslHandshakeTimeoutException: handshake timed out after 10000ms
> * 
> ```

@trask My bad, it's intermittent error. Updated in quote.

**Comment by trask** [comment]
does the problem go away with the latest release (3.4.0 or later)?

**Comment by jiec-msft** [comment]
Haven't upgraded to 3.4.0 or later versions for Azure Spring Apps service, so don't know yet. Let me try check it today.

**Comment by jiec-msft** [comment]
Hi @trask,

I deployed 5 apps with agents at version 3.2.11, 3.3.1, 3.4.0, 3.4.1, 3.4.2. And observed below information.

- I see the error `java.lang.IllegalArgumentException: Failed to load any of the given libraries: [netty_tcnative_linux_x86_64, netty_tcnative_linux_x86_64_fedora, netty_tcnative_x86_64, netty_tcnative]` for below agents, didn't find the error message for 3.2.11.
  - 3.3.1
  - 3.4.0
  - 3.4.1
  - 3.4.2
- This time, I didn't see SSL handshake timeout warning message. But by check Application Insights service instance, Live metrics for agent 3.3.1 looks to be intermittent. (For this one, similar to the one reported in our internal incident)
  - ![image](https://user-images.githubusercontent.com/22190245/198528071-f4d461b9-9728-4672-ba61-cb45aba987cf.png)

- Attach logs in file "ai-logs-1028.zip" with agent self-diagnostic log level at Trace.
  - [ai-logs-1028.zip](https://github.com/microsoft/ApplicationInsights-Java/files/9886125/ai-logs-1028.zip)

**Comment by trask** [comment]
> I see the error java.lang.IllegalArgumentException: Failed to load any of the given libraries: [netty_tcnative_linux_x86_64, netty_tcnative_linux_x86_64_fedora, netty_tcnative_x86_64, netty_tcnative] for below agents

this is expected, it should only be logged at DEBUG level(?)

> Live metrics for agent 3.3.1 looks to be intermittent

do you see this problem with 3.4.0 or later?

**Comment by jiec-msft** [comment]
For Live metrics, it’s not intermittent for agent 3.4.0 or later. These agents are OK.

For error message, you’re right, it is at Debug level.

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Trask Stalnaker ***@***.***>
发送时间: Saturday, October 29, 2022 1:04:28 AM
收件人: microsoft/ApplicationInsights-Java ***@***.***>
抄送: Mason Chen ***@***.***>; Mention ***@***.***>
主题: Re: [microsoft/ApplicationInsights-Java] Observed in debug level message "Failed to load netty-tcnative; OpenSslEngine will be unavailable" (Issue #2640)


I see the error java.lang.IllegalArgumentException: Failed to load any of the given libraries: [netty_tcnative_linux_x86_64, netty_tcnative_linux_x86_64_fedora, netty_tcnative_x86_64, netty_tcnative] for below agents

this is expected, it should only be logged at DEBUG level(?)

Live metrics for agent 3.3.1 looks to be intermittent

do you see this problem with 3.4.0 or later?

―

**Comment by trask** [comment]
> For Live metrics, it’s not intermittent for agent 3.4.0 or later. These agents are OK.

interesting, it may be some other fix that was included in the newer netty version then

**Comment by trask** [comment]
@jiec-msft since this issue appears resolved in 3.4.0 and later, is it ok to close this?

**Comment by jiec-msft** [comment]
@trask Sure, let me close it.
